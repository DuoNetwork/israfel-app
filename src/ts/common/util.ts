import { Constants, IUserOrder, Util as CommonUtil } from '@finbook/israfel-common';
import * as d3 from 'd3';
import moment from 'moment';
import * as CST from './constants';
import { ICustodianInfo } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = CommonUtil.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'just now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' min ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' hrs ago';
		else return 'long time ago';
	}

	public formatExpiry(isWeek: boolean) {
		return moment(CommonUtil.getExpiryTimestamp(isWeek)).format('YYYY-MM-DD HH:mm');
	}

	public formatTime(time: any) {
		return moment(time).format('MM-DD HH:mm:ss');
	}

	public formatTimeSecond(time: number) {
		return moment(time).format('YYYY-MM-DD HH:mm:ss');
	}

	public formatPriceShort(num: number) {
		if (num < 1) return d3.format('.3f')(num);
		return d3
			.format('.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatStrike(num: number) {
		return d3.format('.4f')(num);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatPercent(num: number) {
		return d3.format('.0%')(num);
	}

	public formatPercentAcc(num: number) {
		return d3.format('.2%')(num);
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatMaturity(maturity: number) {
		return maturity ? moment(maturity).format('YYYY-MM-DD HH:mm') : CST.TH_PERPETUAL;
	}

	public getOrderTitle(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order Placed';
		else if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_FILL) return 'Order Fully Filled';
			else if (order.status === Constants.DB_TERMINATE) return 'Order Expired';
			// order.status === Constants.DB_CONFIRMED
			else return 'Order Cancelled';
		else if (order.status === Constants.DB_MATCHING) return 'Order is Being Filled';
		// if (order.status === Constants.DB_PFILL)
		else return 'Order Partially Filled';
	}

	public getOrderFullDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		let baseDescription = `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}.`;
		if (order.type === Constants.DB_ADD)
			return baseDescription + ` Valid till ${this.formatMaturity(order.expiry)}`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_FILL)
			return baseDescription + ' Fully filled';

		if (order.fill || order.matching) {
			const totalFill =
				order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching);
			baseDescription += ` ${totalFill}(${this.formatPercent(
				totalFill / order.amount
			)}) filled.`;
		}

		if (order.matching && order.type !== Constants.DB_TERMINATE)
			baseDescription += 'Pending settlement';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_CONFIRMED)
			return baseDescription + ' Cancelled by user.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_TERMINATE)
			return baseDescription + ' Expired.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_BALANCE)
			return baseDescription + ` Cancelled due to insufficient balance.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_MATCHING)
			return baseDescription + ` Cancelled due to settlement error.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_RESET)
			return baseDescription + ` Cancelled due to custodian reset.`;

		return baseDescription;
	}

	public getOrderDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		return `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}. Total ${util.formatBalance(
			order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching)
		)} filled.`;
	}

	public getVersionDescription(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order placed.';

		if (order.type === Constants.DB_UPDATE) {
			let description = `Total ${this.formatBalance(order.fill + order.matching)} filled`;
			if (order.matching) description += ', pending settlement';
			return description;
		}

		if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_TERMINATE) return 'Expired';
			else if (order.status === Constants.DB_BALANCE)
				return 'Cancelled due to insufficent balance';
			else if (order.status === Constants.DB_MATCHING)
				return 'Cancelled due to settlement error';
			else if (order.status === Constants.DB_RESET) return 'Cancelled due to custodian reset';
			else if (order.status === Constants.DB_FILL) return 'Fully filled';
			// (order.status === Constants.DB_CONFIRMED)
			else return 'Cancelled by user';
		return 'Invalid order';
	}

	public getMaturityDescription(info?: ICustodianInfo) {
		return `This contract ${
			info && info.code.endsWith('PPT')
				? 'never expires'
				: 'expires in ' +
				(info
					? Math.floor(
							(info.states.maturity - CommonUtil.getUTCNowTimestamp()) / 86400000
						).toFixed(0)
					: 0) +
				' days'
		}.`;
	}

	public convertOrdersToCSV(orders: { [pair: string]: IUserOrder[] }): string {
		const headers = [
			CST.TH_PAIR,
			CST.TH_ORDER_HASH,
			CST.TH_VERSION,
			CST.TH_TIME_UTC,
			CST.TH_SIDE,
			CST.TH_PRICE,
			CST.TH_AMOUNT,
			CST.TH_FILL,
			CST.TH_FEE,
			CST.TH_EXPIRY_UTC,
			CST.TH_STATUS,
			CST.TH_TX_HASH
		];
		let output = '';
		headers.forEach(
			(header, index) =>
				(output += this.wrapCSVString(header) + (index < headers.length - 1 ? ',' : '\n'))
		);
		const ordersByHash: { [orderHash: string]: IUserOrder[] } = {};
		for (const pair in orders)
			orders[pair].forEach(order => {
				if (!ordersByHash[order.orderHash]) ordersByHash[order.orderHash] = [];
				ordersByHash[order.orderHash].push(order);
			});

		const rows: any[][] = [];
		for (const orderHash in ordersByHash) {
			ordersByHash[orderHash].sort((a, b) => a.currentSequence - b.currentSequence);
			ordersByHash[orderHash].forEach((order, index) => {
				const row: any[] = [];
				row.push(order.createdAt);
				row.push(order.pair.replace('|', '-')); // CST.TH_PAIR,
				row.push(order.orderHash); // CST.TH_ID,
				row.push(index + 1); // CST.TH_VERSION,
				row.push(
					moment.utc(order.updatedAt || order.createdAt).format('YYYY-MM-DD HH:mm:ss')
				); // CST.TH_TIME_UTC,
				row.push(order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL); // CST.TH_SIDE,
				row.push(order.price); // CST.TH_PRICE,
				row.push(order.amount); // CST.TH_AMOUNT,
				row.push(order.fill + order.matching); // CST.TH_FILL,
				row.push(order.fee + ' ' + order.feeAsset); // CST.TH_FEE,
				row.push(moment.utc(order.expiry).format('YYYY-MM-DD HH:mm:ss')); // CST.TH_EXPIRY,
				// CST.TH_STATUS,
				if (order.type === Constants.DB_TERMINATE)
					if (order.status === Constants.DB_BALANCE)
						row.push('Cancelled due to insufficent balance');
					else if (order.status === Constants.DB_MATCHING)
						row.push('Cancelled due to settlement error');
					else if (order.status === Constants.DB_RESET)
						row.push('Cancelled due to custodian reset');
					else if (order.status === Constants.DB_FILL) row.push('Fully filled');
					else if (order.status === Constants.DB_CONFIRMED) row.push('Cancelled by user');
					else row.push('Expired');
				else if (order.status === Constants.DB_MATCHING) row.push('Pending settlement');
				else row.push('Open');
				row.push(order.transactionHash || ''); // CST.TH_TX_HASH
				rows.push(row);
			});
		}

		rows.sort((a, b) => a[1].localeCompare(b[1]) || -a[0] + b[0] || -a[3] + b[3]);
		rows.forEach(row =>
			row.forEach((item, index) => {
				if (index)
					output +=
						this.wrapCSVString(item + '') + (index !== row.length - 1 ? ',' : '\n');
			})
		);

		return output;
	}

	public wrapCSVString(input: string): string {
		return ('' + input).indexOf(',') >= 0 ? '"' + input + '"' : input;
	}

	private getEtherScanLink() {
		return `https://${__ENV__ === Constants.DB_LIVE ? '' : 'kovan.'}etherscan.io`;
	}

	public getEtherScanTransactionLink(txHash: string) {
		return `${this.getEtherScanLink()}/tx/${txHash}`;
	}

	public getEtherScanTokenLink(tokenAddr: string, account: string = '') {
		return `${this.getEtherScanLink()}/token/${tokenAddr}${account ? `?a=${account}` : ''}`;
	}

	public getEtherScanAddressLink(account: string) {
		return `${this.getEtherScanLink()}/address/${account}`;
	}

	public fillZero(num: number) {
		return num < 10 ? `0${num}` : `${num}`;
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}
}

const util = new Util();
export default util;
