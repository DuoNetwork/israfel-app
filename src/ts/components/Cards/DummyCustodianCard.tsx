import { Constants as WrapperConstants } from '@finbook/duo-contract-wrapper';
import { Icon } from 'antd';
import * as React from 'react';
import * as CST from 'ts/common/constants';
import { ColorStyles } from 'ts/common/styles';
import PriceChart from 'ts/components/Charts/PriceChart';
import { SDivFlexCenter } from '../_styled';
import { SButton, SCard, SCardList, SCardTitle } from './_styled';

interface IProps {
	type: string;
	margin: string;
}

export default class CustodianCard extends React.Component<IProps> {
	public render() {
		const { margin, type } = this.props;
		const isBeethoven = type === WrapperConstants.BEETHOVEN;
		const aLabel = '-';
		const bLabel = '-';
		return (
			<SCard
				title={
					<SCardTitle style={{ cursor: 'Pointer' }}>{type + ' Loading...'}</SCardTitle>
				}
				width="360px"
				margin={margin}
			>
				<SDivFlexCenter horizontal>
					<SCardList noLiBorder noUlBorder noMargin width="56%">
						<div className="status-list-wrapper">
							<ul>
								<li>
									<span
										className="title"
										style={{ display: 'flex', alignItems: 'center' }}
									>
										{CST.TH_ETH}
									</span>
									<span
										className="title"
										style={{
											display: 'flex',
											alignItems: 'center',
											transform: 'scaleX(1.5)'
										}}
									>
										<Icon type="swap" />
									</span>
									<span className="content">
										{isBeethoven ? (
											<div className="cus-des">
												<div>Income Token</div>
												<div>Leverage Token</div>
											</div>
										) : (
											<div className="cus-des">
												<div>Short Token</div>
												<div>Long Token</div>
											</div>
										)}
									</span>
								</li>
								<li style={{ justifyContent: 'start' }}>
									<span className="title" style={{ fontSize: 10 }}>
										Fully backed by
									</span>
									<span
										className="content"
										style={{ fontSize: 10, marginLeft: 4 }}
									>
										<b>{'0' + CST.TH_ETH}</b>
									</span>
								</li>
							</ul>
						</div>
					</SCardList>
					<div
						style={{
							width: '44%',
							display: 'flex',
							justifyContent: 'center',
							alignItems: 'center',
							padding: '0 0 0 10px'
						}}
					>
						<SButton>{CST.TH_CONVERT}</SButton>
					</div>
				</SDivFlexCenter>
				<SDivFlexCenter
					horizontal
					padding="10px 5px 0 5px"
					style={{ color: ColorStyles.TextBlackAlpha, fontSize: 12 }}
				>
					<div className="cuscardtokenwrapper">
						<span style={{ display: 'flex', alignItems: 'center' }}>
							-<span className="aspan">{isBeethoven ? 'INCOME' : 'SHORT'}</span>
						</span>
						<span>{aLabel}</span>
					</div>
					<div className="cuscardnavtag">
						<span className="navspan">NAV</span>
						<span>
							<b>{'$ 0'}</b>
						</span>
					</div>
				</SDivFlexCenter>
				<SDivFlexCenter horizontal height="90px" padding="5px 0 15px 0">
					<div style={{ width: '56%' }}>
						<PriceChart prices={[]} timeOffset={7 * 3600} name={'-'} isA={true} />
					</div>
					<div
						style={{
							width: '44%',
							padding: '5px 0 0 10px',
							display: 'flex',
							flexDirection: 'column',
							justifyContent: 'space-between'
						}}
					>
						<div style={{ display: 'flex', justifyContent: 'space-between' }}>
							<div className="px-left">
								<b>-</b>
							</div>
							<div className="px-right">
								<b>-</b>
							</div>
						</div>
						<SButton>{CST.TH_TRADE}</SButton>
					</div>
				</SDivFlexCenter>
				<SDivFlexCenter
					horizontal
					padding="0 5px 0 5px"
					style={{ color: ColorStyles.TextBlackAlpha, fontSize: 12 }}
				>
					<div className="cuscardtokenwrapper">
						<span style={{ display: 'flex', alignItems: 'center' }}>
							-<span className="aspan">LONG</span>
						</span>
						<span>{bLabel}</span>
					</div>
					<div className="cuscardnavtag">
						<span className="navspan">NAV</span>
						<span>
							<b>{'$ 0'}</b>
						</span>
					</div>
				</SDivFlexCenter>
				<SDivFlexCenter horizontal height="80px" padding="5px 0 5px 0">
					<div style={{ width: '56%' }}>
						<PriceChart prices={[]} timeOffset={7 * 3600} name={'-'} isA={false} />
					</div>
					<div
						style={{
							width: '44%',
							padding: '5px 0 0 10px',
							display: 'flex',
							flexDirection: 'column',
							justifyContent: 'space-between'
						}}
					>
						<div style={{ display: 'flex', justifyContent: 'space-between' }}>
							<div className="px-left">
								<b>-</b>
							</div>
							<div className="px-right">
								<b>-</b>
							</div>
						</div>
						<SButton>{CST.TH_TRADE}</SButton>
					</div>
				</SDivFlexCenter>
				<SDivFlexCenter horizontal width="30%" padding="0">
					{[1, 3, 7].map(pct => (
						<SButton
							key={pct + ''}
							className={'range-picker ' + (pct === 7 ? '' : 'day-Button')}
							style={{ border: 'none' }}
						>
							{pct === 1 ? '24H' : pct + 'D'}
						</SButton>
					))}
				</SDivFlexCenter>
			</SCard>
		);
	}
}
