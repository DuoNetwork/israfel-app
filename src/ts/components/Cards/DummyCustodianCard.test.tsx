/**
 * @jest-environment jsdom
 */
// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import { Util as CommonUtil } from '@finbook/israfel-common';
import { shallow } from 'enzyme';
import * as React from 'react';
import util from 'ts/common/util';
import DummyCustodianCard from './DummyCustodianCard';

describe('DummyCustodianCard Test', () => {
	describe('Test Snapshot', () => {
		const type = "test";
		const margin = "test";
		it('Test Snapshot', () => {
			util.formatMaturity = jest.fn(() => '1970-01-01 08:00:00');
			util.formatExpiry = jest.fn(() => '1970-01-01 19:00:00');
			CommonUtil.getUTCNowTimestamp = jest.fn(() => 1234567890);
			const wrapper = shallow(
				<DummyCustodianCard
					type={type}
					margin={margin}
				/>
			);
			expect(wrapper).toMatchSnapshot();
			});
	});
});
