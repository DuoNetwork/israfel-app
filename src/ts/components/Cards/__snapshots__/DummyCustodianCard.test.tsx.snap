// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DummyCustodianCard Test Test Snapshot Test Snapshot 1`] = `
<Styled(Card)
  margin="test"
  title={
    <styled.div
      style={
        Object {
          "cursor": "Pointer",
        }
      }
    >
      test Loading...
    </styled.div>
  }
  width="360px"
>
  <styled.div
    horizontal={true}
  >
    <styled.div
      noLiBorder={true}
      noMargin={true}
      noUlBorder={true}
      width="56%"
    >
      <div
        className="status-list-wrapper"
      >
        <ul>
          <li>
            <span
              className="title"
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                }
              }
            >
              ETH
            </span>
            <span
              className="title"
              style={
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "transform": "scaleX(1.5)",
                }
              }
            >
              <Icon
                type="swap"
              />
            </span>
            <span
              className="content"
            >
              <div
                className="cus-des"
              >
                <div>
                  Short Token
                </div>
                <div>
                  Long Token
                </div>
              </div>
            </span>
          </li>
          <li
            style={
              Object {
                "justifyContent": "start",
              }
            }
          >
            <span
              className="title"
              style={
                Object {
                  "fontSize": 10,
                }
              }
            >
              Fully backed by
            </span>
            <span
              className="content"
              style={
                Object {
                  "fontSize": 10,
                  "marginLeft": 4,
                }
              }
            >
              <b>
                0ETH
              </b>
            </span>
          </li>
        </ul>
      </div>
    </styled.div>
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
          "padding": "0 0 0 10px",
          "width": "44%",
        }
      }
    >
      <styled.button>
        Convert
      </styled.button>
    </div>
  </styled.div>
  <styled.div
    horizontal={true}
    padding="10px 5px 0 5px"
    style={
      Object {
        "color": "rgba(64,79,84,.8)",
        "fontSize": 12,
      }
    }
  >
    <div
      className="cuscardtokenwrapper"
    >
      <span
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        -
        <span
          className="aspan"
        >
          SHORT
        </span>
      </span>
      <span>
        -
      </span>
    </div>
    <div
      className="cuscardnavtag"
    >
      <span
        className="navspan"
      >
        NAV
      </span>
      <span>
        <b>
          $ 0
        </b>
      </span>
    </div>
  </styled.div>
  <styled.div
    height="90px"
    horizontal={true}
    padding="5px 0 15px 0"
  >
    <div
      style={
        Object {
          "width": "56%",
        }
      }
    >
      <TimeSeriesChart
        isA={true}
        name="-"
        prices={Array []}
        timeOffset={25200}
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "space-between",
          "padding": "5px 0 0 10px",
          "width": "44%",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
          }
        }
      >
        <div
          className="px-left"
        >
          <b>
            -
          </b>
        </div>
        <div
          className="px-right"
        >
          <b>
            -
          </b>
        </div>
      </div>
      <styled.button>
        Trade
      </styled.button>
    </div>
  </styled.div>
  <styled.div
    horizontal={true}
    padding="0 5px 0 5px"
    style={
      Object {
        "color": "rgba(64,79,84,.8)",
        "fontSize": 12,
      }
    }
  >
    <div
      className="cuscardtokenwrapper"
    >
      <span
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
          }
        }
      >
        -
        <span
          className="aspan"
        >
          LONG
        </span>
      </span>
      <span>
        -
      </span>
    </div>
    <div
      className="cuscardnavtag"
    >
      <span
        className="navspan"
      >
        NAV
      </span>
      <span>
        <b>
          $ 0
        </b>
      </span>
    </div>
  </styled.div>
  <styled.div
    height="80px"
    horizontal={true}
    padding="5px 0 5px 0"
  >
    <div
      style={
        Object {
          "width": "56%",
        }
      }
    >
      <TimeSeriesChart
        isA={false}
        name="-"
        prices={Array []}
        timeOffset={25200}
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "space-between",
          "padding": "5px 0 0 10px",
          "width": "44%",
        }
      }
    >
      <div
        style={
          Object {
            "display": "flex",
            "justifyContent": "space-between",
          }
        }
      >
        <div
          className="px-left"
        >
          <b>
            -
          </b>
        </div>
        <div
          className="px-right"
        >
          <b>
            -
          </b>
        </div>
      </div>
      <styled.button>
        Trade
      </styled.button>
    </div>
  </styled.div>
  <styled.div
    horizontal={true}
    padding="0"
    width="30%"
  >
    <styled.button
      className="range-picker day-Button"
      key="1"
      style={
        Object {
          "border": "none",
        }
      }
    >
      24H
    </styled.button>
    <styled.button
      className="range-picker day-Button"
      key="3"
      style={
        Object {
          "border": "none",
        }
      }
    >
      3D
    </styled.button>
    <styled.button
      className="range-picker "
      key="7"
      style={
        Object {
          "border": "none",
        }
      }
    >
      7D
    </styled.button>
  </styled.div>
</Styled(Card)>
`;
